### ISSUE TITLE
Optimizing a custom loss in TF1 vs TF2

### ISSUE THREAD

**Issue by osotsia** [tensorflow/tensorflow#45186]
I'm trying to migrate from TF1 to TF2 and encountering performance issues. The plan is to optimize a custom loss function (without neural networks), but it is much faster on TF1 compared to TF2. Am I doing something wrong?

Below I tried to include a minimum reproducible example. On Colab TF1 takes 0.061 seconds, while TF2 takes 0.277 seconds.

[TF1 colab notebook](https://colab.research.google.com/drive/1H0bPxIrSvD9V0IHppNpeQRwzHP4qbal9?usp=sharing)

[TF2 colab notebook](https://colab.research.google.com/drive/1tZ9aHXc9EA6otTyRIVBFhK8ez3yjXaA2?usp=sharing)

**Comment by amahendrakar** [comment]
Was able to reproduce the issue. Running the code with [TF v1.15](https://colab.research.google.com/gist/amahendrakar/5cc50d9204aee1439972b3a54af56707/45186-1-15.ipynb#scrollTo=q4UqYm-4qcWH) takes 0.12 seconds. 

Whereas, running with [TF v2.3](https://colab.research.google.com/gist/amahendrakar/2563b517066b7b8acfa32944a1dc87e5/45186-2-x.ipynb#scrollTo=XPz3p-ubqldR) and [TF-nightly](https://colab.research.google.com/gist/amahendrakar/75014533bda0cb2ef1ab1d4dd807262b/45186-2-5.ipynb#scrollTo=XPz3p-ubqldR) takes 0.52 and 0.41 seconds respectively. Please find the attached gist. Thanks!

**Comment by amahendrakar** [comment]
@osotsia, 
The Colab notebook you've linked contains the custom loss, back propagation and optimizer. Could you isolate the performance differences into one of them, for example only custom loss itself? Having an xprof of TF1 vs TF2 would also help analysis. Thanks.

**Comment by osotsia** [comment]
@amahendrakar
Please be more clear. Are you asking me to find out exactly where the TF2 performance hit is coming from (loss, backprop, or adagrad)? Any suggestions where to begin? I suppose it's easy to benchmark just the loss function (and it's unlikely to be the issue), but for the other two I'm not at all familiar with the TF2 code base. I don't know what xpof is either, and [google](https://www.google.com/search?channel=fs&q=xpof+tensorflow) is not very helpful.

**Comment by rmothukuru** [comment]
@osotsia,
It looks like you have written equivalent code of `Tensorflow Version 1.x`, in `2.x`. Can you migrate your `1.x` code to `2.x` using [Upgrade Script](https://www.tensorflow.org/guide/upgrade) and [Migration Guide](https://www.tensorflow.org/guide/migrate) and see if the issue persists. Thanks!

**Comment by osotsia** [comment]
@rmothukuru
Hi, thanks for that. It works without the performance hit. If this is a long-term solution then I'm happy to close this issue.

**Comment by rmothukuru** [comment]
@osotsia,
Yes, following Upgrade Script and Migration Guide is the recommended if we have to migrate the code for `TF Version` from `1.x` to `2.x`.  Thanks!
